wi-fi/hal/cyhal_sdio.c

```diff
@@ -4,766 +4,128 @@
-				cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "SDIO cmd error (Command End Bit Error)\n");
-			if (val & (1 << 17)) /* CCE=1 */
-				cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "SDIO cmd error (Command CRC Error)\n");
-			if (val & (1 << 16)) /* CTOE=1 */
-				cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "SDIO cmd error (Command Timeout Error)\n");
-			return -1;
-		}
-
-		if ((val & flags) == flags)
-			return 0;
-
-		usleep(wait_us);
-	}
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_WARNING, "SDIO cmd timeout\n");
-
-	return -1;
-}
-
-
-static int reset_all(uint8_t sdclkfs, uint8_t dvs)
-{
-	uint32_t i, val;
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO\n");
-
-	*(sdio_common.base + sys_ctrl) = (1 << 24) | (sdclkfs << 8) | (dvs << 4) | 0xf; /* RSTA=1 */
-
-	// FIXME: ???
-	usleep(10);
-
-	for (i = 0; i < 10; ++i) {
-		val = *(sdio_common.base + sys_ctrl) & (1 << 24);
-		if (val == 0)
-			return 0;
-		usleep(10);
-	}
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO\n");
-
-	return -1;
-}
-
-
-static int reset_cmd_block(void)
-{
-	uint32_t i, val;
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO cmd block\n");
-
-	val = *(sdio_common.base + sys_ctrl);
-	val |= 1 << 25; /* RSTC=1 */
-	*(sdio_common.base + sys_ctrl) = val;
-
-	// FIXME: ???
-	usleep(10);
-
-	for (i = 0; i < 10; ++i) {
-		val = *(sdio_common.base + sys_ctrl) & (1 << 25);
-		if (val == 0)
-			return 0;
-		usleep(10);
-	}
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO cmd block\n");
-
-	return -1;
-}
 
+#include <sdio.h>
 
-static int reset_data_block(void)
-{
-	uint32_t i, val;
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_DEBUG, "reset SDIO data block\n");
-
-	val = *(sdio_common.base + sys_ctrl);
-	val |= 1 << 26; /* RSTD=1 */
-	*(sdio_common.base + sys_ctrl) = val;
-
-	// FIXME: ???
-	usleep(10);
-
-	for (i = 0; i < 10; ++i) {
-		val = *(sdio_common.base + sys_ctrl) & (1 << 26);
-		if (val == 0)
-			return 0;
-		usleep(10);
-	}
-
-	cy_log_msg(CYLF_SDIO, CY_LOG_ERR, "failed to reset SDIO data block\n");
-
-	return -1;
-}
-
-
-static void sdio_irq_thread(void *arg)
-{
-	bool finish;
-	uint32_t val;
 
-	while (1) {
-		mutexLock(sdio_common.irq_lock);
+typedef union {
+	struct {
+		uint8_t data;                /* 0 - 7 */
+		unsigned int _stuff2 : 1;    /* 8     */
+		unsigned int address : 17;   /* 9-25  */
+		unsigned int _stuff : 1;     /* 26    */
+		unsigned int checkWrite : 1; /* 27    */
+		unsigned int areaIndex : 3;  /* 28-30 */
+		unsigned int dir : 1;        /* 31    */
+	} cmd52;
+	struct {
+		unsigned int count : 9;     /* 0-8   */
+		unsigned int address : 17;  /* 9-25  */
+		unsigned int opCode : 1;    /* 26    */
+		unsigned int blockMode : 1; /* 27    */
+		unsigned int areaIndex : 3; /* 28-30 */
+		unsigned int dir : 1;       /* 31    */
+	} cmd53;
+	uint32_t uint;
+} sdio_cmd_arg_t;
 
-		while (1) {
-			finish = sdio_common.irq_thread_finish;
-			if (finish)
-				break;
 
-			val = *(sdio_common.base + int_status);
-			if (val & (1 << 8))
-				break;
+cyhal_sdio_irq_handler_t conv;
```

Comment:
At least it should be static. But I think the most portable solution would be to add a structure for keeping the sdio context with only this value here and make use of the (cyhal_sdio_t *obj) pointer, which gets passed to basically every function.